avm-tf-modules-review-doc (#988)
Azure/Azure-Verified-Modules: docs/content/contributing/terraform/review.md

---

> title: AVM Review of Terraform Modules

Reviewer: ```suggestion
title: Initial Review of Terraform Modules
```

Author: Would like to keep the original title, we have plans to further split this into two sections namely "New Module Reviews" and "Existing Module Reviews" in future

geekdocNav: true
geekdocAlign: left
geekdocAnchor: true
---

{{< toc >}}

> AVM Review is a critical step before an AVM terraform module is published to [Terraform Registry](https://registry.terraform.io/) and made publicly available for customers, partners and wider IaC community to consume and contribute to. It acts as a quality assurance step to ensure that the AVM terraform module complies with the [shared](https://azure.github.io/Azure-Verified-Modules/specs/shared/) and [terraform](https://azure.github.io/Azure-Verified-Modules/specs/terraform/) specifications of AVM. The below process outlines the steps for both a module owner and a module reviewer should follow to do an AVM Review of a terraform module.

Reviewer: ```suggestion
The initial AVM module review is a critical step before an AVM terraform module gets published to the [Terraform Registry](https://registry.terraform.io/) and made publicly available for customers, partners and wider IaC community to consume and contribute to. It serves as a quality assurance step to ensure that the AVM terraform module complies with the [shared](https://azure.github.io/Azure-Verified-Modules/specs/shared/) and [terraform](https://azure.github.io/Azure-Verified-Modules/specs/terraform/) specifications of AVM. The below process outlines the steps that both the module owner and module reviewer need to follow.
```

1. The module owner completes the development of the module in their branch or fork.

> 2. The module owner does a pull request (PR) titled 'AVM-Review-PR' ensuring that all checks are passing on that PR as that is a pre-requisite to request a review.

Reviewer: ```suggestion
2. The module owner submits a pull request (PR) titled 'AVM-Review-PR' and ensures that all checks are passing on that PR as that is a pre-requisite to request a review.
```

> 3. The module owner assigns the `avm-core-team-technical-terraform` as reviewers on the PR.

Reviewer: ```suggestion
3. The module owner assigns the `avm-core-team-technical-terraform` GitHub team as reviewer on the PR.
```

> 4. The module owner leaves the following comment as is on the module proposal in the AVM Repo by searching for their module proposal here.

Reviewer: ```suggestion
4. The module owner leaves the following comment as is on the module proposal in the [AVM Repo](https://aka.ms/avm/moduleproposals) by searching for their module proposal issue there.
```

<br>"I would like to request a review of my module. The latest code is in a PR titled AVM-Review-PR on the module repo and all checks on that PR are passing."

> 5. The AVM team will move the module proposal from "In Development" to "In Review" in the [AVM - Module Triage](https://github.com/orgs/Azure/projects/529) project.

Reviewer: ```suggestion
5. The AVM team moves the module proposal from "In Development" to "In Review" in the [AVM - Module Triage](https://github.com/orgs/Azure/projects/529) project.
```

> 6. The AVM team will assign a module reviewer who will open a blank issue on the module titled "AVM-Review" and populate it with the below mark down. This template already marks the specs as compliant which are covered by the checks that run on the PR. There are some specs which don't need to be checked at the time of publishing the module therefore they are marked as NA.

Reviewer: @mbilalamjad, I really like this flow, but why don't we add this as an issue template through the template repo? It could just be called "Initial module review" - this way people won't have to copy text around.

Author: That's a good idea but worried that it might confuse module owners and module consumers as to when to create an issue with that template

{{< expand "➕ AVM Terraform Module Review Issue" "expand/collapse" >}}

{{< include file="static/includes/avm-terraform-module-review-template.md" language="markdown" options="linenos=false" >}}

{{< /expand >}}

> 7. The module reviewer can update the Compliance column for specs 42 to 47 to NA in case the module being reviewed isn't a pattern module.

Reviewer: Please reference the full spec ID instead of just the 2 digit number from their end.

Reviewer: ```suggestion
7. The module reviewer can update the Compliance column for specs 42 to 47 to NA, in case the module being reviewed is a resource module.
```

> 8. The module reviewer to review the code in the PR and make comments to request module owner of any necessary updates to the PR.

Reviewer: ```suggestion
8. The module reviewer will review the code in the PR and make comments to request any necessary updates.
```

> 9. The module reviewer will assign the AVM-Review issue to the module owner and respond to the module owner's comment on the Module Proposal in AVM Repo with the following

Reviewer: ```suggestion
9. The module reviewer assigns the AVM-Review issue to the module owner and responds to the module owner's comment on the Module Proposal in AVM Repo with the following
```

{{< expand "➕ AVM Terraform Module Review Initiation Message" "expand/collapse" >}}

> "Thank you for reaching out to request a review of your module. The AVM module review process has been initiated, kindly perform the **Requested Actions** on the AVM-Review issue on the module repo."

Reviewer: ```suggestion
"Thank you for requesting a review of your module. The AVM module review process has been initiated, please perform the **Requested Actions** on the AVM-Review issue on the module repo."
```

> "Thank you for reaching out to request a review of your module. The AVM module review process has been initiated, kindly perform the **Requested Actions** on the AVM-Review issue on the module repo."

Reviewer 2: try to make use of 
``` 
text
```
or `<!-- markdownlint-disable -->` to let the text pop out it would be nice so people's eyes don't get bored or loose attention reading through the text because it's already quite some text.

Reviewer 2: But it's just a suggestion or nice-to-have ;)

Author: Thanks for the suggestion, have incorporated it.

{{< /expand >}}

> 10. The module owner will update the check list and the table in the AVM-Review issue and accordingly update the module reviewer.

Reviewer: ```suggestion
10. The module owner updates the check list and the table in the AVM-Review issue and notifies the module reviewer in a comment.
```

> 11. The module reviewer will perform the final review and ensure that all checks in the checklist are complete and the specs table has been updated with no spec's having compliance as 'No'.

Reviewer: ```suggestion
11. The module reviewer will perform the final review and ensure that all checks in the checklist are complete and the specifications table has been updated with no requirements having compliance as 'No'.
```

> 12. The module reviewer will approve the AVM-Review-PR and make the following comment on the AVM-Review issue and close it.

Reviewer: ```suggestion
12. The module reviewer approves the AVM-Review-PR and leaves the following comment on the AVM-Review issue and close it.
```

{{< expand "➕ AVM Terraform Module Review Completion Message" "expand/collapse" >}}

> Thank you so much for contributing this module and completing the review process per AVM specs to ensure quality.

Reviewer: ```suggestion
Thank you for contributing this module and completing the review process per AVM specs to ensure quality.
```

You may proceed with publishing this module to the HashiCorp Terraform Registry with the initial pre-release version of v0.1.0.

> Please keep future versions also pre-release (e.g., 0.1.0, 0.1.1, 0.2.0, etc.) until AVM goes to GA of which the AVM team will notify.

Reviewer: ```suggestion
Please keep future versions also pre-release (e.g., 0.1.0, 0.1.1, 0.2.0, etc.) until AVM becomes generally available (GA) of which the AVM team will notify you.
```

> **Requested Action**: Once the module is published to the HashiCorp Terraform Registry kindly update the proposal with the below comment.

Reviewer: ```suggestion
**Requested Action**: Once the module is published to the HashiCorp Terraform Registry, please update the modyule proposal issue with the below comment.
```

> "The AVM Review of this module has been completed and the module has been published to the registry. Requesting AVM team to close this module proposal and update module index accordingly."

Reviewer: ```suggestion
"The initial review of this module has been completed and the module has been published to the registry. Requesting AVM team to close this module proposal and mark the module available in the module index."
```

{{< /expand >}}

> 13. The module owner will merge the AVM-Review-PR and publish the module to Terraform registry and update the module proposal on the AVM repo with the following

Reviewer: ```suggestion
13. The module owner will merge the AVM-Review-PR, publish the module to Terraform registry and update the module proposal on the AVM repo with the following
```

{{< expand "➕ AVM Terraform Module Publish Message" "expand/collapse" >}}

> The AVM Review of this module has been completed and the module has been published to the registry. Requesting AVM team to close this module proposal and update module index accordingly.

Reviewer: ```suggestion
The initial review of this module is complete and the module has been published to the registry. Requesting AVM team to close this module proposal and mark the module available in the module index.
```

{{< /expand >}}

> 14. The module reviewer will update the module proposal by doing the following

Reviewer: ```suggestion
14. The module reviewer updates the module proposal by performing the following steps
```

- Move the module proposal from "In Review" to "Done" in the [AVM - Module Triage](https://github.com/orgs/Azure/projects/529) project.
- Assign label Status: Module Available :green_circle:
- Update the module index

> - Close the module proposal

Reviewer: @mbilalamjad, where do we check if the mandatory GH teams have been created with the required permissions granted?

Author: Its one of the specs in the table which we ask the module owner to confirm even though we have automation to check it
